[mlir][Interfaces] `LoopLikeOpInterface`: Support ops with multiple regions

matthias-springer opened the pull request:
This commit implements `LoopLikeOpInterface` on `scf.while`. This enables LICM (and potentially other transforms) on `scf.while`.

`LoopLikeOpInterface::getLoopBody()` is renamed to `getLoopRegions` and can now return multiple regions.

Also fix a bug in the default implementation of `LoopLikeOpInterface::isDefinedOutsideOfLoop()`, which returned "false" for some values that are defined outside of the loop (in a nested op, in such a way that the value does not dominate the loop). This interface is currently only used for LICM and there is no way to trigger this bug, so no test is added.

zero9178 commented on mlir/test/Transforms/loop-invariant-code-motion.mlir:
```diff
@@ -929,3 +929,24 @@ func.func @speculate_dynamic_pack_and_unpack(%source: tensor<?x?xf32>,
+// -----
+
+// CHECK-LABEL: func @hoist_from_scf_while(
+//   CHECK-DAG:   arith.constant 1 : i32
+//   CHECK-DAG:   arith.constant 10 : i32
+//       CHECK:   scf.while
+//   CHECK-NOT:     arith.constant
```
Not sure whether its worth changing, but hoisting `arith.constant` isn't really interesting for LICM given that `canonicalize` is capable of hoisting them as well. Maybe more complicated pure ops like `arith.cmpi` or `arith.addi` should be tested instead.

zero9178 commented on mlir/include/mlir/Interfaces/LoopLikeInterface.td:
```diff
@@ -36,14 +36,14 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
       }],
-      /*retTy=*/"::mlir::Region &",
+      /*retTy=*/"::llvm::SmallVector<::mlir::Region *>",
       /*methodName=*/"getLoopBody"
```
Is it worth renaming this to `getLoopRegions` or similar now? I feel like "body" has somewhat of a consistent meaning in MLIR, so far referring to a possibly single block region of a structured op, while in the new signature it now as the more abstract meaning of "the regions that make up the loop".

zero9178 reviewed (approved):
I added two comments that are rather bikesheddy, so feel free to ignore them if you don't see the point in them. 
Otherwise, LGTM as this is a mostly mechanical change.

Do note that there is a seemingly real build failure in Flang according to CI.

matthias-springer committed:
[mlir][Interfaces] `LoopLikeOpInterface`: Support ops with multiple regions

This commit implements `LoopLikeOpInterface` on `scf.while`. This enables LICM (and potentially other transforms) on `scf.while`.

`LoopLikeOpInterface::getLoopBody()` can now return multiple regions.

Also fix a bug in the default implementation of `LoopLikeOpInterface::isDefinedOutsideOfLoop()`, which returned "false" for some values that are defined outside of the loop (in a nested op, in such a way that the value does not dominate the loop). This interface is currently only used for LICM and there is no way to trigger this bug, so no test is added.

BEGIN_PUBLIC
No public commit message needed for presubmit.
END_PUBLIC
